test: support std::optional in BOOST_CHECK_* and increase FromUserHex fuzz feature coverage

l0rinc opened the pull request:
Enhanced `FromUserHex` coverage by:

* Added `std::optional` support to `BOOST_CHECK_EQUAL`, allowing direct comparisons of `std::optional<T>` with other `T` expected values.
* Increased fuzz testing for hex parsing to validate against other hex validators and parsers.

----

* Use BOOST_CHECK_EQUAL for https://github.com/bitcoin/bitcoin/pull/30569#discussion_r1706637780 arith_uint256, uint256, uint160

Example error before:
> unknown location:0: fatal error: in "validation_chainstatemanager_tests/chainstatemanager_args": std::bad_optional_access: bad_optional_access
test/validation_chainstatemanager_tests.cpp:781: last checkpoint

after:
> test/validation_chainstatemanager_tests.cpp:801: error: in "validation_chainstatemanager_tests/chainstatemanager_args": check set_opts({"-assumevalid=0"}).assumed_valid_block == uint256::ZERO has failed [std::nullopt != 0000000000000000000000000000000000000000000000000000000000000000]

l0rinc committed:
Compare FromUserHex result against other hex validators and parsers

hodlinator commented on src/test/util/setup_common.h:
```diff
@@ -39,13 +40,42 @@ extern const std::function<std::vector<const char*>()> G_TEST_COMMAND_LINE_ARGUM
+    }
+}
+
+// Enable BOOST_CHECK_EQUAL for arith_uint256
+inline std::ostream& operator<<(std::ostream& os, const arith_uint256& num)
+{
+    return os << num.ToString();
+}
+
+// Enable BOOST_CHECK_EQUAL for uint256
+inline std::ostream& operator<<(std::ostream& os, const uint256& num)
+{
+    return os << num.ToString();
+}
+
+// Enable BOOST_CHECK_EQUAL for uint160
+inline std::ostream& operator<<(std::ostream& os, const uint160& num)
+{
+    return os << num.ToString();
+}
```
Why:
1. Put these 3 inside `namespace std`?
2. `inline` these non-templated functions?
   I know one day we will have modules and be happy without .H/CPP distinction, but until then, let us try to keep headers lean. Especially when the coding style mandates minimum 4 lines per function implementation, and an #include needs adding.

hodlinator reviewed (commented):
Concept ACK a42afbdcfe02b2a872fe9bfa6cd98287b1fa9a50

TIL: one can compare `optional` against a value without `.value()` (C++17 addition).

<details>
<summary>
Would suggest something closer to this:
</summary>

```diff
diff --git a/src/test/util/setup_common.h b/src/test/util/setup_common.h
index 6c1b2a57f1..d1eae29003 100644
--- a/src/test/util/setup_common.h
+++ b/src/test/util/setup_common.h
@@ -5,7 +5,6 @@
 #ifndef BITCOIN_TEST_UTIL_SETUP_COMMON_H
 #define BITCOIN_TEST_UTIL_SETUP_COMMON_H
 
-#include <arith_uint256.h>
 #include <common/args.h> // IWYU pragma: export
 #include <kernel/context.h>
 #include <key.h>
@@ -40,44 +39,6 @@ extern const std::function<std::vector<const char*>()> G_TEST_COMMAND_LINE_ARGUM
 /** Retrieve the unit test name. */
 extern const std::function<std::string()> G_TEST_GET_FULL_NAME;
 
-namespace std {
-// Enable BOOST_CHECK_EQUAL for enum class types
-template <typename T>
-inline std::ostream& operator<<(typename std::enable_if<std::is_enum<T>::value, std::ostream>::type& stream, const T& e)
-{
-    return stream << static_cast<typename std::underlying_type<T>::type>(e);
-}
-
-// Enable BOOST_CHECK_EQUAL for std::optional
-template <typename T>
-inline std::ostream& operator<<(std::ostream& os, const std::optional<T>& v)
-{
-    if (v) {
-        return os << v.value();
-    } else {
-        return os << "std::nullopt";
-    }
-}
-
-// Enable BOOST_CHECK_EQUAL for arith_uint256
-inline std::ostream& operator<<(std::ostream& os, const arith_uint256& num)
-{
-    return os << num.ToString();
-}
-
-// Enable BOOST_CHECK_EQUAL for uint256
-inline std::ostream& operator<<(std::ostream& os, const uint256& num)
-{
-    return os << num.ToString();
-}
-
-// Enable BOOST_CHECK_EQUAL for uint160
-inline std::ostream& operator<<(std::ostream& os, const uint160& num)
-{
-    return os << num.ToString();
-}
-} // namespace std
-
 static constexpr CAmount CENT{1000000};
 
 struct TestOpts {
@@ -280,6 +241,36 @@ std::unique_ptr<T> MakeNoLogFileContext(const ChainType chain_type = ChainType::
 
 CBlock getBlock13b8a();
 
+// Make types usable in BOOST_CHECK_* {
+namespace std {
+// Enum class types
+template <typename T>
+inline std::ostream& operator<<(typename std::enable_if<std::is_enum<T>::value, std::ostream>::type& stream, const T& e)
+{
+    return stream << static_cast<typename std::underlying_type<T>::type>(e);
+}
+
+inline std::ostream& operator<<(std::ostream& os, const std::nullopt_t)
+{
+    return os << "std::nullopt";
+}
+
+template <typename T>
+inline std::ostream& operator<<(std::ostream& os, const std::optional<T>& v)
+{
+    if (v) {
+        return os << v.value();
+    } else {
+        return os << std::nullopt;
+    }
+}
+} // namespace std
+
+std::ostream& operator<<(std::ostream& os, const arith_uint256& num);
+std::ostream& operator<<(std::ostream& os, const uint256& num);
+std::ostream& operator<<(std::ostream& os, const uint160& num);
+// Make types usable in BOOST_CHECK_* }
+
 /**
  * BOOST_CHECK_EXCEPTION predicates to check the specific validation error.
  * Use as
```
</details>

l0rinc commented on src/test/util/setup_common.h:
```diff
@@ -39,13 +40,42 @@ extern const std::function<std::vector<const char*>()> G_TEST_COMMAND_LINE_ARGUM
+    if (v) {
+        return os << v.value();
+    } else {
+        return os << "std::nullopt";
```
This seemed like a simplification to me, what would be the advantage of that?

l0rinc replied:
Done, but I don't yet see what this adds exactly

hodlinator replied:
Being able to write `BOOST_CHECK_EQUALS(..., std::nullopt);`, in case one wanted to.

l0rinc replied:
but we're never writing optionals on the right side here, i.e.
```C++
BOOST_CHECK_EQUAL(uint256::FromUserHex(""), std::make_optional(uint256::ZERO));
```
can simply be either:
```C++
BOOST_CHECK_EQUAL(uint256::FromUserHex(""), uint256::ZERO);
```
or if empty:
```C++
BOOST_CHECK(!uint256::FromUserHex(""));
```

l0rinc replied:
Removed the `nullopt` overwrite and explained the usage in the comment, thanks for explaining your view.

stickies-v replied:
I think I agree that until we have use case (I can't think of one atm) to compare with `std::nullopt`, forcing the user to use `BOOST_CHECK` makes sense.

In commit e0a96acc82c2c918572a2c9b90431db5bca19174:
> Note that we're never writing optionals on the right side, e.g.
> BOOST_CHECK_EQUAL(uint256::FromUserHex(""), std::make_optional(uint256::ZERO));
since they can simply be either:
> BOOST_CHECK_EQUAL(uint256::FromUserHex(""), uint256::ZERO);
or if empty:
> BOOST_CHECK(!uint256::FromUserHex(""));

I think this can be dropped from the commit message? It's possible, and there are good use cases for putting an optional on the right side, e.g. [here](https://github.com/stickies-v/bitcoin/commit/1f2b0fa86db2ed65476b68417aa1bf4c26026a19#diff-fc0f86b6c3bb23b0e983bcf79d7546d1c9eaa15d6e4d8a7b03b5b85955f585abR682-R684)). You could add a note to the developer notes on this, but I don't think that's necessary either.

l0rinc replied:
Wouldn't
```C++
        BOOST_CHECK_EQUAL(from_hex.has_value(), parse_hash_v.has_value());
        if (!from_hex.has_value()) continue;
        BOOST_CHECK_EQUAL(from_hex.value(), parse_hash_v.value());
```
be simpler after this change as just directly comparing the two?
```C++
        BOOST_CHECK_EQUAL(from_hex, parse_hash_v);
```
?

stickies-v replied:
Yes, in which case we'd be putting an optional (variable) on the right hand side, which goes against what you're saying in the commit message. That's why I suggest removing it, it doesn't help with reviewing this commit imo. It only adds context on why it doesn't do something that was done in another commit (that's not in the tree).

l0rinc replied:
> I think this can be dropped from the commit message

Done, your right, the formulation wasn't the best

l0rinc committed:
Add std::optional support to Boost's equality check

Also moved the operators to the bottom of the file since they're less important and to group them together.

Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>
Co-authored-by: stickies-v <stickies-v@protonmail.com>

l0rinc committed:
Use BOOST_CHECK_EQUAL for optional, arith_uint256, uint256, uint160

Example error before:
> unknown location:0: fatal error: in "validation_chainstatemanager_tests/chainstatemanager_args": std::bad_optional_access: bad_optional_access
test/validation_chainstatemanager_tests.cpp:781: last checkpoint

after:
> test/validation_chainstatemanager_tests.cpp:801: error: in "validation_chainstatemanager_tests/chainstatemanager_args": check set_opts({"-assumevalid=0"}).assumed_valid_block == uint256::ZERO has failed [std::nullopt != 0000000000000000000000000000000000000000000000000000000000000000]

Also added extra minimum_chainwork test to make it symmetric with assumevalid

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>
Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

l0rinc commented on src/test/util/setup_common.h:
```diff
@@ -250,11 +280,6 @@ std::unique_ptr<T> MakeNoLogFileContext(const ChainType chain_type = ChainType::
-// Make types usable in BOOST_CHECK_*
-std::ostream& operator<<(std::ostream& os, const arith_uint256& num);
-std::ostream& operator<<(std::ostream& os, const uint160& num);
-std::ostream& operator<<(std::ostream& os, const uint256& num);
```
Done

l0rinc commented:
> TIL: one can compare `optional` against a value without `.value()` (C++17 addition).

Yeah, but I don't think that's used here

hodlinator commented:
> > TIL: one can compare `optional` against a value without `.value()` (C++17 addition).
>
> Yeah, but I don't think that's used here

It is used in your change from
```C++
BOOST_CHECK_EQUAL(uint256::FromUserHex("").value(), uint256::ZERO);
```
to
```C++
BOOST_CHECK_EQUAL(uint256::FromUserHex(""), uint256::ZERO);
``` 
It was my obscure way of saying thank you for making me discover that. :)

l0rinc commented:
> It is used in your change from
BOOST_CHECK_EQUAL(uint256::FromUserHex("").value(), uint256::ZERO);

I don't think it's related, commenting out:
```diff
-inline std::ostream& operator<<(std::ostream& os, const std::nullopt_t)
-{
-    return os << "std::nullopt";
-}
-
-template <typename T>
-inline std::ostream& operator<<(std::ostream& os, const std::optional<T>& v)
-{
-    if (v) {
-        return os << v.value();
-    } else {
-        return os << std::nullopt;
-    }
-}
+//inline std::ostream& operator<<(std::ostream& os, const std::nullopt_t)
+//{
+//    return os << "std::nullopt";
+//}
+//
+//template <typename T>
+//inline std::ostream& operator<<(std::ostream& os, const std::optional<T>& v)
+//{
+//    if (v) {
+//        return os << v.value();
+//    } else {
+//        return os << std::nullopt;
+//    }
+//}
```
and running
> cmake -B build && cmake --build build -j10 && build/src/test/test_bitcoin --run_test=uint256_tests/from_user_hex
gives me:
```bash
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk/usr/include/c++/v1/ostream:268:20: note: candidate function not viable: no known conversion from 'const std::optional<arith_uint256>' to 'nullptr_t' (aka 'std::nullptr_t') for 1st argument
    basic_ostream& operator<<(nullptr_t)
                   ^
4 errors generated.
```

hodlinator commented:
Maybe my communication was not clear enough. There are 2 separate things that you might be conflating?

1. I like that one can compare `optional<T>` to `T` as per https://en.cppreference.com/w/cpp/utility/optional/operator_cmp (21-33).
    That's the `BOOST_CHECK_EQUAL(uint256::FromUserHex(""), uint256::ZERO);`-example.

2. I don't see why we want to prevent being able to do:
    `BOOST_CHECK_EQUAL(uint256::FromUserHex("Not valid"), std::nullopt);`
    one could do the following as an alternative, but `BOOST_CHECK` is deprecated:
    `BOOST_CHECK(!uint256::FromUserHex("Not valid"));`

Not sure where your `nullptr_t` error is coming from. I misspelled `std::nullopt` as `nullptr` once while attempting to verify #2.

l0rinc commented:
> 2\. but `BOOST_CHECK` is deprecated

I don't see the deprecation https://live.boost.org/doc/libs/1_86_0/libs/test/doc/html/boost_test/utf_reference/testing_tool_ref/assertion_boost_level.html, I think it's still useful for truthy checks (e.g. empty optionals).

> one can compare optional against a value without .value() (C++17 addition).

Yes, you're right here, I wrongly assumed some BOOST magic was done in this case instead.

hodlinator commented:
> > 2. but `BOOST_CHECK` is deprecated
> 
> I don't [see the deprecation](https://live.boost.org/doc/libs/1_86_0/libs/test/doc/html/boost_test/utf_reference/testing_tool_ref/assertion_boost_level.html), I think it's still useful for truthy checks (e.g. empty optionals).

It seems to be poorly documented, but if you search [the header](https://www.boost.org/doc/libs/1_86_0/boost/test/tools/interface.hpp) for "DEPRECATED TOOLS" you'll find it.

stickies-v commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -22,15 +20,14 @@
 #include <validation.h>
 #include <validationinterface.h>
 
-#include <tinyformat.h>
```
What's the rationale behind these include updates? According to `iwyu` this include is still necessary. More generally, I think it's most productive to either fix all the includes, or the ones that are affected by the PR, but not a (seemingly random) subset?

stickies-v commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -806,22 +803,24 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_args, BasicTestingSetup)
+    BOOST_CHECK_EQUAL(get_valid_opts({"-assumevalid="}).assumed_valid_block, uint256::ZERO);
+    BOOST_CHECK_EQUAL(get_valid_opts({"-assumevalid=0"}).assumed_valid_block, uint256::ZERO);
+    BOOST_CHECK_EQUAL(get_valid_opts({"-noassumevalid"}).assumed_valid_block, uint256::ZERO);
+    BOOST_CHECK_EQUAL(get_valid_opts({"-assumevalid=0x1234"}).assumed_valid_block, uint256::FromHex("0000000000000000000000000000000000000000000000000000000000001234"));
```
nit: could use the consteval `uint256` hex string ctor here (+ below)

```suggestion
    BOOST_CHECK_EQUAL(get_valid_opts({"-assumevalid=0x1234"}).assumed_valid_block, uint256{"0000000000000000000000000000000000000000000000000000000000001234"});
```



<details>
<summary>git diff on f509948e1e</summary>

```diff
diff --git a/src/test/validation_chainstatemanager_tests.cpp b/src/test/validation_chainstatemanager_tests.cpp
index 3914deb926..e9d9fee1fe 100644
--- a/src/test/validation_chainstatemanager_tests.cpp
+++ b/src/test/validation_chainstatemanager_tests.cpp
@@ -807,7 +807,7 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_args, BasicTestingSetup)
     BOOST_CHECK_EQUAL(get_valid_opts({"-assumevalid="}).assumed_valid_block, uint256::ZERO);
     BOOST_CHECK_EQUAL(get_valid_opts({"-assumevalid=0"}).assumed_valid_block, uint256::ZERO);
     BOOST_CHECK_EQUAL(get_valid_opts({"-noassumevalid"}).assumed_valid_block, uint256::ZERO);
-    BOOST_CHECK_EQUAL(get_valid_opts({"-assumevalid=0x1234"}).assumed_valid_block, uint256::FromHex("0000000000000000000000000000000000000000000000000000000000001234"));
+    BOOST_CHECK_EQUAL(get_valid_opts({"-assumevalid=0x1234"}).assumed_valid_block, uint256{"0000000000000000000000000000000000000000000000000000000000001234"});
     auto assume_valid{"0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"s};
     BOOST_CHECK_EQUAL(get_valid_opts({("-assumevalid=" + assume_valid).c_str()}).assumed_valid_block, uint256::FromHex(assume_valid));
 
@@ -818,7 +818,7 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_args, BasicTestingSetup)
     BOOST_CHECK(!get_valid_opts({}).minimum_chain_work);
     BOOST_CHECK_EQUAL(get_valid_opts({"-minimumchainwork=0"}).minimum_chain_work, arith_uint256());
     BOOST_CHECK_EQUAL(get_valid_opts({"-nominimumchainwork"}).minimum_chain_work, arith_uint256());
-    BOOST_CHECK_EQUAL(get_valid_opts({"-minimumchainwork=0x1234"}).minimum_chain_work, UintToArith256(*uint256::FromHex("0000000000000000000000000000000000000000000000000000000000001234")));
+    BOOST_CHECK_EQUAL(get_valid_opts({"-minimumchainwork=0x1234"}).minimum_chain_work, UintToArith256(uint256{"0000000000000000000000000000000000000000000000000000000000001234"}));
     auto minimum_chainwork{"0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"s};
     BOOST_CHECK_EQUAL(get_valid_opts({("-minimumchainwork=" + minimum_chainwork).c_str()}).minimum_chain_work, UintToArith256(*uint256::FromHex(minimum_chainwork)));
 

```
</details>

stickies-v commented on src/test/uint256_tests.cpp:
```diff
@@ -395,16 +397,16 @@ BOOST_AUTO_TEST_CASE(from_hex)
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("10"), uint256{0x10});
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("0xFf"), uint256{0xff});
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("Ff"), uint256{0xff});
+    auto valid_hex_64{"0x0123456789abcdef0123456789abcdef0123456789ABDCEF0123456789ABCDEF"s};
```
I think it's best to keep the `const` qualifier? (and less importantly, also [here](https://github.com/bitcoin/bitcoin/pull/30618/files#diff-dbada1fe3a3d0af884304dd28be8c9df74b592401dec2c6400f6b491aefe6c9bR811) and [here](https://github.com/bitcoin/bitcoin/pull/30618/files#diff-dbada1fe3a3d0af884304dd28be8c9df74b592401dec2c6400f6b491aefe6c9bR822))

l0rinc replied:
I don't really see the value in doing that for such short-scoped test variables - but you can convince me with good arguments, it just seems like noise to me in these cases

stickies-v replied:
I don't think `valid_hex_64` is so short-scoped? Sure, you can manually validate that the 9 places where it's used don't mutate it, but why not make it explicit and fool-proof with `const`?

l0rinc replied:
Fair, added const

stickies-v commented on src/test/fuzz/hex.cpp:
```diff
@@ -34,8 +32,14 @@ FUZZ_TARGET(hex)
+        assert(result_string.length() == 64);
+        assert(IsHex(result_string));
+        assert(TryParseHex(result_string));
+        assert(Txid::FromHex(result_string));
+        assert(Wtxid::FromHex(result_string));
+        assert(uint256::FromHex(result_string));
```
I'm not sure these 3 lines are worth adding, since they're already covered [higher up](https://github.com/bitcoin/bitcoin/pull/30618/files#diff-12dd31b274739b6a132d9c57e5ab14b403436991e4201cf8d3f3cc4fc7995066R29-R34), so this might unnecessarily slow down the test?

l0rinc replied:
those were testing the same input, this one is testing the parsing of `result->ToString()`

stickies-v replied:
Isn't that just a very expensive way to do `assert(random_hex_string == ToLower(result->ToString())`? (and also, that's not `hex` testing, but `uint256` testing)

l0rinc replied:
isn't the point of these tests that we shouldn't have know the internals (i.e. black box)?
(though in this case that wouldn't even be true since `random_hex_string` can start with `0x` for `FromUserHex` which isn't added back via `ToString`)

stickies-v replied:
> (though in this case that wouldn't even be true since `random_hex_string` can start with `0x` for `FromUserHex` which isn't added back via `ToString`)

Ah yes good point thanks, perhaps that would be good to clarify in a docstring?

> isn't the point of these tests that we shouldn't have know the internals (i.e. black box)?

I'm out of my depth here, so will probably refrain from commenting further and hopefully someone with more fuzzing experience can chime in. I think the point of the fuzz tests is to find bugs, and my understanding is that we often tweak them (e.g. based on knowledge about the implementation) to make them more effective at doing that. As such, not testing the `{Txid,Wtxid,uint256}::FromHex()` asserts twice seems like a reasonable improvement, but I'm not sure.

marcofleon replied:
On my machine, running

`FUZZ=hex build_fuzz/src/test/fuzz/fuzz ../qa-assets/fuzz_seed_corpus/hex -runs=1000000`

gives 681 seconds with those three lines and 533 seconds without them. Exec/s are reasonable (>1000) for both versions of the harness. The slow down isn't too drastic, so if these added checks are useful in any way then I think it's fine to keep them.

stickies-v replied:
Thanks for looking into it @marcofleon! If there are no operational concerns adding the test cases, I'm happy with it.

I guess if we keep this, the `uint256::FromHex()` asserts have become superfluous and could be removed. Shouldn't make a meaningful performance impact (since they're called much less often than the `FromUserHex()` asserts, so it's more of a thing to keep the test code lean.

<details>
<summary>git diff on 6fa4c92e4c</summary>

```diff
diff --git a/src/test/fuzz/hex.cpp b/src/test/fuzz/hex.cpp
index 1109186a8f..71518f626a 100644
--- a/src/test/fuzz/hex.cpp
+++ b/src/test/fuzz/hex.cpp
@@ -28,12 +28,7 @@ FUZZ_TARGET(hex)
     if (IsHex(random_hex_string)) {
         assert(ToLower(random_hex_string) == hex_data);
     }
-    if (uint256::FromHex(random_hex_string)) {
-        assert(random_hex_string.length() == 64);
-        assert(Txid::FromHex(random_hex_string));
-        assert(Wtxid::FromHex(random_hex_string));
-        assert(uint256::FromUserHex(random_hex_string));
-    }
+    (void)uint256::FromHex(random_hex_string);
     if (auto result{uint256::FromUserHex(random_hex_string)}) {
         const auto result_string{result->ToString()};
         assert(result_string.length() == 64);

```
</details>

l0rinc replied:
I don’t fully understand why we’re focusing on optimization at this stage. Why not keep both options for now, and if we later find that performance is an issue, remove the redundant ones based on their actual implementations? In my opinion, we’re still in the discovery phase, where having as much redundancy as we can afford is beneficial. Since fuzzing—and property-based testing in general—is black-box testing, we acknowledge that we don’t know the full input space. Our goal is to ensure that certain properties are upheld, even though we might not fully understand how those properties could be violated.

l0rinc replied:
> We seem to disagree

We don't have to, if the above discussion didn't change your mind I'll gladly remove them.

l0rinc replied:
Done

stickies-v replied:
I'm hesitant to add more to the back-and-forth, but I think we may have been talking past each other. In my [latest comment](https://github.com/bitcoin/bitcoin/pull/30618#discussion_r1744124233) I suggested updating the `FromHex()` tests, where based on the [latest force-push](https://github.com/bitcoin/bitcoin/compare/6fa4c92e4c9bf5ea90e25d41b525c88f9c2d834d..2cee181a6cc00840def16c311aae69883130eebd) it seems you were talking about the `FromUserHex()` tests, which I agreed I was okay with based on @marcofleon's [testing](https://github.com/bitcoin/bitcoin/pull/30618#discussion_r1743946931).

@marcofleon  also showed me (and I agree with) that removing `assert(uint256::FromHex(result_string));` is a slight regression since it's not a duplicate (and [I specifically added it](https://github.com/bitcoin/bitcoin/pull/30569/commits/70e2c87737e77ee85812cc328c4ddfaea7147533#diff-12dd31b274739b6a132d9c57e5ab14b403436991e4201cf8d3f3cc4fc7995066R38)), so that could be added back in? Sorry for not highlighting that earlier, I thought we'd moved past keeping all the `FromUserHex()` tests already so I didn't want to add noise.

l0rinc replied:
> I'm hesitant to add more to the back-and-forth

I don't mind jousting until we find something that is better than what either of us had in mind. Happens often, keep insisting on parts that you feel strongly about!
Let's see what you think of my latest push:
* FromHex correctness assumes FromUserHex parseability
* FromUserHex correctness assumes that if reconverted to hex, it should be accepted by FromHex

Pushed (first a separate rebase, and a change)

stickies-v replied:
> Let's see what you think of my latest push:

Yeah I like the current state of it, thanks for addressing this and can be marked as resolved now!

stickies-v commented on src/test/fuzz/hex.cpp:
```diff
@@ -34,8 +32,14 @@ FUZZ_TARGET(hex)
-    if (const auto result{uint256::FromUserHex(random_hex_string)}) {
-        assert(uint256::FromHex(result->ToString()));
+    if (auto result{uint256::FromUserHex(random_hex_string)}) {
+        auto result_string = result->ToString();
```
nit
```suggestion
        const auto result_string{result->ToString()};
```

l0rinc replied:
Done (without the const part, which just seems like noise in this context to me, but am open to changing my mind about it)

stickies-v replied:
Again, here, I don't think the `const` is noise. This variable is passed to multiple functions, and verifying no mutation happens is an unnecessary burden on the developer. Why not just lean on the compiler?

l0rinc replied:
K, I don't feel strongly about it, pushed.

stickies-v reviewed (commented):
Approach ACK 50defb82c5c52f9cca00de53f76064c376220329

Being able to compare `std::optional` directly in `BOOST_CHECK_*` macros is very helpful, thank you for picking this up. The added fuzz coverage seems helpful too but I'm lacking fuzzing experience to comment in detail here.

A couple of suggested improvements besides the comments left:
1. The PR description would benefit from a bit more context/rationale on what this PR does and why it is helpful (`TrySanitizeHexNumber` is also no longer a thing).
1. The example in the f509948e1e17d244c40e3e3b38c755ec5b5f5d08 commit message is helpful but outdated (`set_opts` doesn't exist)
1. I think the PR title is not very accurate, would suggest e.g. `test: support std::optional in BOOST_CHECK_* and increase hex fuzz coverage` (we're not increasing `FromUserHex` fuzz coverage imo)

l0rinc commented on src/test/util/setup_common.h:
```diff
@@ -250,10 +241,30 @@ std::unique_ptr<T> MakeNoLogFileContext(const ChainType chain_type = ChainType::
+
+template <typename T>
+inline std::ostream& operator<<(std::ostream& os, const std::optional<T>& v)
+{
+    if (v) {
+        return os << v.value();
+    } else {
+        return os << "std::nullopt";
+    }
+}
```
Yeah, I tried that before, but getting:
> error: incompatible operand types ('const value_type' (aka 'const uint256') and 'const char[13]')
    return os << v ? *v : "std::nullopt";

stickies-v replied:
Sorry you're right, `return v ? os << *v : os << "std::nullopt";` should work though.

l0rinc replied:
Use the version in my edited comment above

stickies-v replied:
I don't think requiring `v` to have implemented `ToString()` is a good approach. Do you see any issues with `return v ? os << *v : os << "std::nullopt";`?

l0rinc replied:
I'm fine with that as well, pushed.

l0rinc commented on src/test/util/setup_common.cpp:
```diff
@@ -79,24 +73,6 @@ const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
 /** Random context to get unique temp data dirs. Separate from m_rng, which can be seeded from a const env var */
 static FastRandomContext g_rng_temp_path;
 
-std::ostream& operator<<(std::ostream& os, const arith_uint256& num)
```
was specifically requested to avoid repetition and since they're not mission-critical code: https://github.com/bitcoin/bitcoin/pull/30618#discussion_r1739923996

l0rinc replied:
Pushed some changes, let me know if it's any better, or if you insist on placing them in the cpp file in a different place than the header (with some explanations)

stickies-v replied:
I don't insist on placing them in a different place than in the header file, I just don't think this is worth the noise and bigger diff - I don't think anyone uses a declaration's location in the header to find the implementation in the cpp file?

l0rinc replied:
I think it makes it slightly more aligned with the header file, while the diff isn't a lot heavier.
What do you think @hodlinator?

hodlinator replied:
Was under the impression that Git had some smarts when it came to moving lines around, but at least naive `git blame` usage shows your name & commit on all the moved lines in the **.cpp**. It's straightforward code (going back in history is less necessary), and I understand your attempt to mirror the ordering from the **.h** file.

I think the ordering I pushed for in the **.h** file should stay either way, but maybe go with the more conservative approach in the **.cpp** as a compromise.

stickies-v replied:
No I don't feel strongly about it, I just think this change is not nearly useful enough to warrant an extra 33LoC diff. Anyway, not worth arguing over, like I said it's not blocking.

l0rinc commented on src/test/util/setup_common.h:
```diff
@@ -250,10 +241,30 @@ std::unique_ptr<T> MakeNoLogFileContext(const ChainType chain_type = ChainType::
+}
+
+template <typename T>
+inline std::ostream& operator<<(std::ostream& os, const std::optional<T>& v)
```
The iwyu changes were a bit excessive, but kept the ones that are related to this PR, thanks!

l0rinc commented on src/test/util/setup_common.h:
```diff
@@ -250,10 +241,30 @@ std::unique_ptr<T> MakeNoLogFileContext(const ChainType chain_type = ChainType::
 CBlock getBlock13b8a();
 
-// Make types usable in BOOST_CHECK_*
+// Make types usable in BOOST_CHECK_* {
```
Used the `@{` and `@}` formats

hodlinator replied:
`@{` is [Doxygen grouping syntax](https://www.doxygen.nl/manual/grouping.html). I was trying to make it work properly in generated documentation using `@name` but the `namespace std` parts were always being extracted, so opted for vanilla comment style.

l0rinc commented on src/test/util/setup_common.h:
```diff
@@ -250,10 +241,30 @@ std::unique_ptr<T> MakeNoLogFileContext(const ChainType chain_type = ChainType::
+inline std::ostream& operator<<(typename std::enable_if<std::is_enum<T>::value, std::ostream>::type& stream, const T& e)
+{
+    return stream << static_cast<typename std::underlying_type<T>::type>(e);
+}
```
nice! changed

l0rinc commented:
Thanks for the thorough review @hodlinator and @stickies-v, I pushed another batch (first an empty rebase), followed by https://github.com/bitcoin/bitcoin/compare/c6f9add458bb8e142d8ccfe5c661ec56143e3f31..6fa4c92e4c9bf5ea90e25d41b525c88f9c2d834d

stickies-v commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -806,22 +807,24 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_args, BasicTestingSetup)
+    BOOST_CHECK(!get_valid_opts({}).minimum_chain_work);
+    BOOST_CHECK_EQUAL(get_valid_opts({"-minimumchainwork=0"}).minimum_chain_work, arith_uint256());
+    BOOST_CHECK_EQUAL(get_valid_opts({"-nominimumchainwork"}).minimum_chain_work, arith_uint256());
+    BOOST_CHECK_EQUAL(get_valid_opts({"-minimumchainwork=0x1234"}).minimum_chain_work, UintToArith256(uint256{"0000000000000000000000000000000000000000000000000000000000001234"}));
```
simplification nit
```suggestion
    BOOST_CHECK_EQUAL(get_valid_opts({"-minimumchainwork=0x1234"}).minimum_chain_work, arith_uint256{0x1234});
```

l0rinc replied:
nice, if I retouch I'll apply it!

stickies-v reviewed (approved):
ACK 6fa4c92e4c9bf5ea90e25d41b525c88f9c2d834d , modulo my lack of fuzzing experience to have a good view on whether the `fuzz/hex.cpp` changes indeed make the fuzz test better (the code looks correct and safe to me), as per https://github.com/bitcoin/bitcoin/pull/30618/files#r1742175853

I [maintain](https://github.com/bitcoin/bitcoin/pull/30618/files#r1742093199) that the `setup_common.cpp` changes should be dropped from dec011bdaaeb000a9f22a31e3fe9415ec7bf97dc, but it is not a blocker.

l0rinc reviewed (commented):
> good point thanks, perhaps that would be good to clarify in a docstring?

Do you think it's important in this case? We're not using the `random_hex_string` as input here at all, only the parsed output, reconverted to hex.
If you do, please give me a diff and I'll apply it.

stickies-v commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -806,22 +807,24 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_args, BasicTestingSetup)
+    BOOST_CHECK_EQUAL(get_valid_opts({"-nominimumchainwork"}).minimum_chain_work, arith_uint256());
+    BOOST_CHECK_EQUAL(get_valid_opts({"-minimumchainwork=0x1234"}).minimum_chain_work, arith_uint256{0x1234});
+    auto minimum_chainwork{"0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"s};
+    BOOST_CHECK_EQUAL(get_valid_opts({("-minimumchainwork=" + minimum_chainwork).c_str()}).minimum_chain_work, UintToArith256(*uint256::FromHex(minimum_chainwork)));
```
Dereferencing could be UB when `FromHex()` is buggy, so for robustness would use `.value()` here.
```suggestion
    BOOST_CHECK_EQUAL(get_valid_opts({("-minimumchainwork=" + minimum_chainwork).c_str()}).minimum_chain_work, UintToArith256(uint256::FromHex(minimum_chainwork).value()));
```

l0rinc replied:
Thanks, fixed

stickies-v reviewed (approved):
ACK 3437d4e1bdec7963d3b49bcc396838602a5b00fc

> Do you think it's important in this case?

No it's fine as-is too, it's just that we start from a string so producing another, seemingly similar one might be unintuitive to someone less familiar with the code. Could add something like this, but no strong view

<details>
<summary>git diff on 3437d4e1bd</summary>

```diff
diff --git a/src/test/fuzz/hex.cpp b/src/test/fuzz/hex.cpp
index 6a85fd9fdb..6a4a381588 100644
--- a/src/test/fuzz/hex.cpp
+++ b/src/test/fuzz/hex.cpp
@@ -32,6 +32,7 @@ FUZZ_TARGET(hex)
         assert(uint256::FromUserHex(random_hex_string));
     }
     if (const auto result{uint256::FromUserHex(random_hex_string)}) {
+        // ToString() returns a fixed-length string without "0x" prefix
         const auto result_string{result->ToString()};
         assert(result_string.length() == 64);
         assert(IsHex(result_string));

```
</details>

l0rinc commented:
> Could add something like this, but no strong view

Done, thanks!

stickies-v commented:
re-ACK 4ac1981d338f89ed2b814aedaebe81783293e6f5

stickies-v commented:
rebase re-ACK 6da47f6ba5450e7f2b2ab2a64c7cfd354f67a1a9, CI timeout failure seems unrelated

ryanofsky commented on src/test/fuzz/hex.cpp:
```diff
@@ -29,13 +29,16 @@ FUZZ_TARGET(hex)
         assert(ToLower(random_hex_string) == hex_data);
     }
     if (uint256::FromHex(random_hex_string)) {
-        assert(random_hex_string.length() == 64);
-        assert(Txid::FromHex(random_hex_string));
-        assert(Wtxid::FromHex(random_hex_string));
```
In commit "Compare FromUserHex result against other hex validators and parsers" (6da47f6ba5450e7f2b2ab2a64c7cfd354f67a1a9)

The three lines removed here are not equivalent to the similar lines added below because these are checking `random_hex_string`, while lines below are checking `result->ToString()`, these strings are not the same because `random_hex_string` can have mixed case. Would suggest not dropping these lines.

stickies-v replied:
This change was made because of [my suggestion](https://github.com/bitcoin/bitcoin/pull/30618#discussion_r1742175853) but you're right there can be a case difference which I did not consider. Sorry for that @l0rinc .

l0rinc replied:
[Done](https://github.com/bitcoin/bitcoin/compare/6da47f6ba5450e7f2b2ab2a64c7cfd354f67a1a9..1eac96a503b6bac3eaf5d0eb3d23ffde3bfbf9aa), thanks both for the suggestions (this is how we get to the best outcome).

ryanofsky reviewed (approved):
Code review ACK 6da47f6ba5450e7f2b2ab2a64c7cfd354f67a1a9. Nice change to make tests easier to read and setup_common code better organized

l0rinc commented on src/test/uint256_tests.cpp:
```diff
@@ -395,16 +397,16 @@ BOOST_AUTO_TEST_CASE(from_hex)
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("10"), uint256{0x10});
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("0xFf"), uint256{0xff});
+    BOOST_CHECK_EQUAL(uint256::FromUserHex("Ff"), uint256{0xff});
+    const auto valid_hex_64{"0x0123456789abcdef0123456789abcdef0123456789ABDCEF0123456789ABCDEF"s};
```
I'm full of surprises - [done](https://github.com/bitcoin/bitcoin/compare/6da47f6ba5450e7f2b2ab2a64c7cfd354f67a1a9..1eac96a503b6bac3eaf5d0eb3d23ffde3bfbf9aa)

l0rinc commented on src/test/validation_chainstatemanager_tests.cpp:
```diff
@@ -806,22 +807,26 @@ BOOST_FIXTURE_TEST_CASE(chainstatemanager_args, BasicTestingSetup)
+    BOOST_CHECK_EQUAL(get_valid_opts({"-noassumevalid"}).assumed_valid_block, uint256::ZERO);
+    BOOST_CHECK_EQUAL(get_valid_opts({"-assumevalid=0x12"}).assumed_valid_block, uint256{0x12});
+
+    auto assume_valid{"0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"s};
```
[Done](https://github.com/bitcoin/bitcoin/compare/6da47f6ba5450e7f2b2ab2a64c7cfd354f67a1a9..1eac96a503b6bac3eaf5d0eb3d23ffde3bfbf9aa)

l0rinc replied:
I didn't `const` them since they're only used in the next line (i.e. very narrow scoped), seems like noise in this case - but if you insist, I'll add it

ryanofsky reviewed (approved):
Code review ACK 1eac96a503b6bac3eaf5d0eb3d23ffde3bfbf9aa. Only changes since last review were auto type and fuzz test tweaks.

hodlinator reviewed (approved):
ACK 1eac96a503b6bac3eaf5d0eb3d23ffde3bfbf9aa

Provides better error messages for tests.

nit: 1eac96a503b6bac3eaf5d0eb3d23ffde3bfbf9aa - "Compare FromUserHex result against other hex validators and parsers" is not dependent on the other changes, so could be moved first instead of last if you re-touch.

stickies-v commented:
re-ACK 1eac96a503b6bac3eaf5d0eb3d23ffde3bfbf9aa

CI timeout issues seem unrelated.
